tests/test_docopt.py

```diff
@@ -708,22 +715,226 @@ def test_issue_71_double_dash_is_not_a_valid_option_argument():
+            Option("-g", None, 1, "gval"),
+            Option("-h", None, 0, False),
+            Option(None, "--india", 0, False),
+            Option("-j", None, 1, "jval"),
+            Option(None, "--k", 1, "kval"),
+        ],
+    ),
+]
+option_examples = [(dedent(doc), options) for (doc, options) in option_examples]
+
+
+@pytest.mark.parametrize("descriptions, options", option_examples)
+def test_parse_options(descriptions, options):
+    assert parse_options(descriptions) == options
+
+
+@pytest.mark.parametrize(
+    "before",
+    [
+        pytest.param("", id="empty"),
+        pytest.param("This is a prog\n", id="1line"),
+        pytest.param("This is a prog\n\nInfo:\n Blah blah\n", id="preceding_sections"),
+    ],
+)
+@pytest.mark.parametrize(
+    "header",
+    [
+        pytest.param("usage:", id="simple"),
+        pytest.param("uSaGe:", id="odd_case"),
+        pytest.param("My Program's Usage:", id="long"),
+        pytest.param("  Indented Usage:", id="indented"),
+    ],
+)
+@pytest.mark.parametrize(
+    "body",
+    [
+        pytest.param("prog [options]", id="simple"),
+        pytest.param(" prog [options]", id="space_simple"),
+        pytest.param("\tprog [options]", id="tab_simple"),
+        pytest.param(" \t prog [options]", id="WS_simple"),
+        pytest.param("\n prog [options]", id="LF_simple"),
+        pytest.param("\n prog [options]\n", id="LF_simple_LF"),
+        pytest.param("prog [options] cmd1\n prog [options] cmd2\n", id="multiple_LF"),
+        pytest.param("\n prog [options] cmd1\n prog [options] cmd2", id="LF_multiple"),
+        pytest.param(
+            "\n prog [options] cmd1\n prog [options] cmd2\n", id="LF_multiple_LF"
+        ),
+        pytest.param(
+            """\
+ prog [options] cmd1
+   [--foo --bar]
+   [--baz --boz]
+ prog [options] cmd2
+""",
+            id="wrapped_arguments",
+        ),
+    ],
+)
+@pytest.mark.parametrize(
+    "after",
+    [
+        pytest.param("", id="empty"),
+        pytest.param("This can be\nany content.\n", id="text"),
+        pytest.param("Options: -a  All", id="single_line"),
+    ],
+)
+def test_parse_docstring_sections(before: str, header: str, body: str, after: str):
+    if after and not body.endswith("\n"):
+        body = body + "\n"
+    assert parse_docstring_sections(before + header + body + after) == (
+        (before, header, body, after)
+    )
+
+
+@pytest.mark.parametrize(
+    "invalid_docstring",
+    [
+        "",
+        # No usage: heading
+        """\
+        Program summary
+
+            myprog [options]
+
+        Options:
+            --foo
+            --bar
+        """,
+        # blank line after usage: heading
+        """\
+        Program summary
+
+        Usage:
+
+            myprog [options]
+
+        Options:
+            --foo
+            --bar
+        """,
+    ],
+)
+def test_parse_docstring_sections__reports_invalid_docstrings(invalid_docstring: str):
```

Comment:
how about a test case for `usage:` being present, but nothing after that?

"""
program summary

usage:
"""

Reply:
Done, that test is now in `test_lint_docstring`.